crates/goose/src/session/session_manager.rs

```diff
@@ -0,0 +1,928 @@
+        Self::instance().await?.get_insights().await
+    }
+
+    pub async fn maybe_update_description(id: &str, provider: Arc<dyn Provider>) -> Result<()> {
+        let session = Self::get_session(id, true).await?;
+        let conversation = session
+            .conversation
+            .ok_or_else(|| anyhow::anyhow!("No messages found"))?;
+
+        let user_message_count = conversation
+            .messages()
+            .iter()
+            .filter(|m| matches!(m.role, Role::User))
+            .count();
+
+        if user_message_count <= MSG_COUNT_FOR_SESSION_NAME_GENERATION {
+            let description = provider.generate_session_name(&conversation).await?;
+            Self::update_session(id)
+                .description(description)
+                .apply()
+                .await
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub struct SessionStorage {
+    pool: Pool<Sqlite>,
+}
+
+pub fn ensure_session_dir() -> Result<PathBuf> {
+    let data_dir = choose_app_strategy(APP_STRATEGY.clone())
+        .expect("goose requires a home dir")
+        .data_dir()
+        .join("sessions");
+
+    if !data_dir.exists() {
+        fs::create_dir_all(&data_dir)?;
+    }
+
+    Ok(data_dir)
+}
+
+fn role_to_string(role: &Role) -> &'static str {
+    match role {
+        Role::User => "user",
+        Role::Assistant => "assistant",
+    }
+}
+
+type SessionRow = (
+    String,
+    String,
+    String,
+    String,
+    String,
+    String,
+    Option<i32>,
+    Option<i32>,
+    Option<i32>,
+    Option<i32>,
+    Option<i32>,
+    Option<i32>,
+    Option<String>,
+    Option<String>,
+);
+
+impl Default for Session {
+    fn default() -> Self {
+        Self {
+            id: String::new(),
+            working_dir: std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")),
+            description: String::new(),
+            created_at: String::new(),
+            updated_at: String::new(),
+            extension_data: ExtensionData::default(),
+            total_tokens: None,
+            input_tokens: None,
+            output_tokens: None,
+            accumulated_total_tokens: None,
+            accumulated_input_tokens: None,
+            accumulated_output_tokens: None,
+            schedule_id: None,
+            recipe: None,
+            conversation: None,
+            message_count: 0,
+        }
+    }
+}
+
+impl Session {
+    pub fn without_messages(mut self) -> Self {
+        self.conversation = None;
+        self
+    }
+}
+
+impl SessionStorage {
+    async fn new() -> Result<Self> {
+        let session_dir = ensure_session_dir()?;
+        let db_path = session_dir.join("sessions.db");
+
+        let storage = if db_path.exists() {
+            Self::open(&db_path).await?
+        } else {
+            println!("Creating new session database...");
+            let storage = Self::create(&db_path).await?;
+
+            if let Err(e) = storage.import_legacy(&session_dir).await {
```

Comment:
Would be good to test this with a few real session dirs from the wild to ensure no issues before launch